sw/device/lib/base/mmio.h

```diff
@@ -234,6 +234,62 @@ inline void mmio_region_nonatomic_set_mask32(mmio_region_t base,
   mmio_region_write32(base, offset, value);
 }
 
+/**
+ * Masked Register field.
+ */
+typedef struct mmio_region_field32 {
+  uint32_t mask;  /**< The field mask. */
+  uint32_t index; /**< The field position within the a register. */
+  uint32_t value; /**< The field value to be written. */
+} mmio_region_field32_t;
+
+/**
+ * Sets the |field| from the MMIO region |base| at the given offset.
+ *
+ * This function performs a non-atomic read-write-modify operation on a
+ * MMIO region.
+ *
+ * @param base the region to set the field in.
+ * @param offset the offset to set the field at, in bytes.
+ * @param field field within selected register field to be set.
+ */
+inline void mmio_region_nonatomic_set_field32(mmio_region_t base,
+                                              ptrdiff_t offset,
+                                              mmio_region_field32_t field) {
+  uint32_t register_value = mmio_region_read32(base, offset);
+  register_value &= ~(field.mask << field.index);
+  register_value |= (field.value & field.mask) << field.index;
+  mmio_region_write32(base, offset, register_value);
+}
+
+/**
+ * Convienience macro for mmio_region_nonatomic_set_fields32.
```

Comment:
This should document how you expect to call it - which I think is with an inline array initialiser containing lots of structs? Something like:
`MMIO_REGION_NONATOMIC_SET_FIELDS32(base, OFFSET_MACRO, [{.mask = …, .index = …, .value = …}, …])` right?

Reply:
Will that even work in C? I thought C had super crazy rules for when an initializer will actually be coerced.

Also, since the commas are not wrapped in parentheses, the last argument needs to be ... to cope with that.

Reply:
My thinking was - creating fields array on stack and passing that to the macro?

Reply:
This code no longer exist.